pkg/pullrequest/scm.go

```diff
@@ -0,0 +1,133 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pullrequest
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+
+	"golang.org/x/oauth2"
+
+	"github.com/jenkins-x/go-scm/scm/driver/github"
+	"github.com/jenkins-x/go-scm/scm/driver/gitlab"
+	"go.uber.org/zap"
+)
+
+func NewSCMHandler(logger *zap.SugaredLogger, raw, provider, token string) (*Handler, error) {
+	u, err := url.Parse(raw)
+	if err != nil {
+		return nil, err
+	}
+
+	if provider == "" {
+		p, err := guessProvider(raw)
+		if err != nil {
+			return nil, err
+		}
+		provider = p
+	}
+
+	var handler *Handler
+	switch provider {
+	case "github":
+		handler, err = githubHandlerFromURL(u, token, logger)
+	case "gitlab":
+		handler, err = gitlabHandlerFromURL(u, token, logger)
+	default:
+		return nil, fmt.Errorf("unsupported pr url: %s", raw)
+	}
+	return handler, err
+}
+
+func githubHandlerFromURL(u *url.URL, token string, logger *zap.SugaredLogger) (*Handler, error) {
+	split := strings.Split(u.Path, "/")
+	if len(split) < 5 {
+		return nil, fmt.Errorf("could not determine PR from URL: %v", u)
+	}
+	owner, repo, pr := split[1], split[2], split[4]
+	prNumber, err := strconv.Atoi(pr)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing PR number: %s", pr)
+	}
+
+	client := github.NewDefault()
+	ownerRepo := fmt.Sprintf("%s/%s", owner, repo)
+	h := NewHandler(logger, client, ownerRepo, prNumber)
+	if token != "" {
+		ts := oauth2.StaticTokenSource(
+			&oauth2.Token{AccessToken: token},
+		)
+		h.client.Client = oauth2.NewClient(context.Background(), ts)
+	}
+	return h, nil
+}
+
+func gitlabHandlerFromURL(u *url.URL, token string, logger *zap.SugaredLogger) (*Handler, error) {
+	// The project name can be multiple /'s deep, so split on / and work from right to left.
+	split := strings.Split(u.Path, "/")
+
+	// The PR number should be the last element.
+	last := len(split) - 1
+	prNum := split[last]
+	prInt, err := strconv.Atoi(prNum)
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse pr as number from %s", u)
+	}
+
+	// Next we sanity check that this is a correct url. The next to last element should be "merge_requests"
+	if split[last-1] != "merge_requests" {
+		return nil, fmt.Errorf("invalid gitlab url: %s", u)
+	}
+
+	// Next, we rejoin everything else into the project field.
+	project := strings.Join(split[1:last-1], "/")
+	client := gitlab.NewDefault()
+	if token != "" {
+		client.Client = &http.Client{
+			Transport: &gitlabClient{
+				token:     token,
+				transport: http.DefaultTransport,
+			},
+		}
+	}
+	return NewHandler(logger, client, project, prInt), nil
+}
+
+// gitlab client wraps a normal http client, adding support for private-token auth.
```

Comment:
Fine for this PR, but this is something that would probably be useful to contribute upstream to the go-scm GitLab driver! ^_^

Reply:
Yeah, I didn't see any examples of auth in the library itself, only in the examples outside. Need to think through that.